_posts/2018-01-09-ble-virtualization.md

```diff
@@ -0,0 +1,125 @@
+---
+title: Bluetooth Low Energy Virtualization
+author: niklasad1
+authors: niklasad1
+---
+
+Hi everyone,
+The Tock team is happy to announce a new feature to the Bluetooth Low Energy
+Advertiser driver which makes is possible for applications to advertise as
+they are different devices.
+This is possible by taking advantage of Process Abstraction in Tock.
+This driver is supported by nRF51-DK and nRF52-DK but because
+nRF51-DK has only 16kB of RAM, there is only memory to run a process at the
+time.
+
+1. TOC
+{:toc}
+
+## Feature description
+Before getting into the actual implementation details we want show you how this
+can be used in Tock. Let's assume you want to run Tock and send Bluetooth Low
+Energy advertisements as four different devices with different configuration
+such as device name, transmitting power, advertisement interval, services and
+similar by using only one chip. This is now possible by running four different
+applications in Tock. To demostrate this for you we run four instances of the
+following application with different configurations:
+
+
+
+```c
+
+#include <simple_ble.h>
+#include <tock.h>
+
+// Sizes in bytes
+#define DEVICE_NAME_SIZE 7
+
+int main(void) {
+  // declarations of variables to be used in this BLE example application
+  uint16_t advertising_interval_ms = 100;
+  uint8_t device_name[] = "TockOS3";
+
+  // configure advertisement interval to 300ms
+  // configure LE only and discoverable
+  ble_initialize(advertising_interval_ms, true);
+
+  // configure device names as TockOS3
+  ble_advertise_name(device_name, DEVICE_NAME_SIZE);
+
+  // start advertising
+  ble_start_advertising();
+
+  return 0;
+}
+
+```
+
```

Comment:
I think this example might be clearer if I'd seen all of the configuration options in the code above, and then knew what I was looking for in the screenshot (rather than just 'things are different for each'), maybe in the code above something like:

```c
if (EXAMPLE_CONFIGURATION == 1) {
  advertising_interval_ms = 100;
  device_name[7] = '1';
} else if (EXAMPLE_CONFIGURATION == 2) {
  advertising_interval_ms = 300;
  device_name[7] = '2';
} ...
```